### PR TITLE
Implement the greedy residual variance minimization algorithm

### PR DESCRIPTION
**Pull request by willtebbutt**
See the references -- I imagine you've come across this already @theogf .

If you have any initial thoughts, that would be great. Otherwise I think I just need to:
- [ ] optimise the partial pivoted cholesky implementation
- [x] agree with you on a name for this method -- I wonder whether it should have the word residual in it somewhere?

### CODE REVIEW

**Comment by theogf** [comment]
Looks great! Could you also add a block in the docs gallery? (in `docs/src/algorithms.md`)

File: src/offline/greedy_var_minimization.jl [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        if d_max < tol * C_max
+            return (V, p, j - 1)
+        end
+
+        if j_max ≢ j
+            switch!(p, j, j_max)
+            switch!(d, j, j_max)
+
+            u .= kernelmatrix(k, x, x[j_max:j_max])
```
Comment (theogf) [inline src/offline/greedy_var_minimization.jl]: You could use view on `x` here as well?

Comment (theogf) [reply src/offline/greedy_var_minimization.jl]: Ah no I just realised it's a one index

**Comment by willtebbutt** [comment]
Regarding the name, I've gone with Greedy Variance Selection, to be in line with what David et al went with.

**Comment by willtebbutt** [comment]
@theogf I think I'm happy with this for now. I might come back and optimise later on, but I'd like to get it in asap if that's okay with you?

**Review by theogf — APPROVED** [review]

File: docs/src/algorithms.md [hunk 2]
```diff
@@ -161,6 +161,27 @@ save("Greedy.svg", fig); nothing # hide
...
+Stops when either `M` pseudo-inputs have been chosen, or `maximum(d) < maximum(diag(C_f)) * tol)`.
+See [Stable and Efficient Gaussian Process Calculations](https://www.jmlr.org/papers/volume10/foster09a/foster09a.pdf) and
+[Convergence of Sparse Variational Inference in Gaussian Processes Regression](https://www.jmlr.org/papers/volume21/19-1015/19-1015.pdf) for a
+discussion of this criterion's use in the sparse GP context.
+In this example, a maximum of 50 inducing points can be used, but the algorithm chooses to use
+less than this as the tolerance is satisfied before all 50 are used.
+
+```@example base
+kernel = with_lengthscale(SqExponentialKernel(), 0.2)
+alg = GreedyVarSelection(50, 1e-6)
```
Comment (theogf) [inline docs/src/algorithms.md:176]: Ah wait sorry a minor comment: could you reuse the `M` parameter. This way it makes the algorithms a bit more comparable to each other.
Although I realize it ruins your point from the text...
If you'd rather keep `50` that's also fine with me.

Comment (theogf) [reply docs/src/algorithms.md:176]: haha never mind

**Pull request merged**